stl/inc/type_traits

```diff
@@ -1657,6 +1657,32 @@ template <class _From, class _To>
 using is_nothrow_convertible = _Is_nothrow_convertible<_From, _To>;
 #endif // _HAS_CXX20
 
+template <class _Ty>
+_Ty _Returns_exactly() noexcept; // not defined
+
+template <class _From, class _To, class = void>
+struct _Invoke_convertible : false_type {};
+
+template <class _From, class _To>
+struct _Invoke_convertible<_From, _To, void_t<decltype(_Implicitly_convert_to<_To>(_Returns_exactly<_From>()))>>
+    : true_type {};
+
+template <class _From, class _To>
+struct _Invoke_nothrow_convertible : bool_constant<noexcept(_Implicitly_convert_to<_To>(_Returns_exactly<_From>()))> {};

```

Comment:
> Am I correct in believing that the `_Implicitly_convertible_to<_To>(_Returns_exactly<_From>())` dance is the major functional change here, and that this is what is addressing the deferred temporary materialization and value category issue mentioned in the [Feedback ticket](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/963790)?

Yes. For `is_invocable_r<R, F, Args...>`, we were previously requiring `is_convertible_v<decltype(invoke(declval<F>(), declval<Args>()...)), R>`, which is equivalent to requiring `_Implicitly_convertible_to<R>(declval<decltype(invoke(declval<F>(), declval<Args>()...))>())` to be well-formed (`_Implicitly_convertible_to<R>` is a function with exactly one parameter of type `R`). This can give the wrong answer when `invoke(declval<F>(), declval<Args>()...)` is a prvalue, because in that case `declval<decltype(invoke(declval<F>(), declval<Args>()...))>()` is an xvalue, and convertibility of prvalues and xvalues of the same type can differ since C++17.

We *could* fix this by spelling out the `invoke` expression explicitly and determining if `_Implicitly_convertible_to<_To>(invoke(declval<F>(), declval<Args>()...))` is well-formed, but we'd like to avoid the cost of evaluating the `invoke` expression again when we can just extract the saved type-and-value-category.

The fix is to define a new function template `_Returns_exactly<T>()` whose return type is exactly `T`, such that `decltype(_Returns_exactly<T>())` is also exactly `T`. We can then ask instead if `_Implicitly_convertible_t<R>(_Returns_exactly<decltype(invoke(declval<F>(), declval<Args>()...))>())` is well-formed which is always the proper question for `is_invocable_r<R, F, Args...>`.

> It looks like neither `_Returns_exactly` or `_Implicitly_convertible_to` are actually defined, so I infer that whether they are well-formed or not is what informs `false_type` or `true_type` here, and I imagine throwing them around as parameters/return types handles the change in value category?

Yes, it's all about the return type of `_Returns_exactly` (which is exactly its template type argument) and the parameter type of `_Implicitly_convertible_to` (which is also exactly its template type argument).

Reply:
> Am I correct in believing that the `_Implicitly_convertible_to<_To>(_Returns_exactly<_From>())` dance is the major functional change here, and that this is what is addressing the deferred temporary materialization and value category issue mentioned in the Feedback ticket?

Yes. The problem was that the old implementation used `is_convertible` (and `_Is_nothrow_convertible`, which uses `is_convertible` internally). That's specified by the Standard in terms of `declval`, and `declval<Object>()` returns `Object&&`. That rvalue reference triggers deferred temporary materialization, so we'll get the wrong answer for `NonMovable` types like in the test. As I understand it, a valid scenario is `void takes_object(NonMovable); NonMovable returns_object(); takes_object(returns_object());`, but if a reference is introduced, `NonMovable&& returns_rvalue_reference(); takes_object(returns_rvalue_reference());` attempts to move-construct `NonMovable` which is doomed.

`_Returns_exactly` is an "improved" `declval` that returns `Object` as a prvalue.

> It looks like neither `_Returns_exactly` or `_Implicitly_convertible_to` are actually defined

Correct - they are imaginary functions like `declval` that exist only for the purposes of `decltype` and `noexcept` expressions.

> so I infer that whether they are well-formed or not is what informs false_type or true_type here

Yes, for `_Invoke_convertible` which uses SFINAE to detect well-formed conversions. `_Invoke_nothrow_convertible` assumes well-formedness, and uses `noexcept` to detect whether the conversion can throw. (The imaginary functions are marked `noexcept` so they cannot be the cause of any exceptions.)

> and I imagine throwing them around as parameters/return types handles the change in value category?

It is specifically the fact that `_Returns_exactly` returns `_Ty` instead of `add_rvalue_reference_t<_Ty>`.

> On another note, any recommendation for getting a better understanding of value categories (other than reading the Standard directly)? 😅

@CaseyCarter recommended [Guaranteed Copy Elision Does Not Elide Copies](https://devblogs.microsoft.com/cppblog/guaranteed-copy-elision-does-not-elide-copies/) by @TartanLlama.

In STL issues/PRs, value categories are relevant whenever we cross over from the expression domain to the type domain (via `decltype`) and back (via `declval` and now `_Returns_exactly`). When we get the `decltype` of an expression, we need to think about its possible value categories because that controls whether it's a reference (and what kind of reference). And when we construct an (imaginary) expression with `declval` etc., we need to make sure that we're forming the right type, cv-qualification, and value category. For example, it is very easy to say `declval<_Ty>()` when what we really need to say is `declval<const _Ty&>()` - for an object type, the former will be "modifiable rvalue of type `_Ty`" and the latter is "`const` lvalue of type `_Ty`" and there can be a significant difference in how they behave.

My techniques for learning are not very surprising - "work on a bunch of such issues/features" and "write an explanation in my own words to see if Casey corrects me" :joy_cat: I wrote some blog posts back in 2010 when I learned rvalue references for the first time but they were lost during the blog migration; I probably still have them in my Sent Items folder but they're pretty seriously outdated by C++17 so I wouldn't recommend them.

Reply:
Thank you both for the detailed explanations!  This is much clearer to me now :)